File: client/py/delta/segment/splitter.py
from . import BinarySegment
from .. import telem


class Splitter:
    threshold: telem.Size

    def __init__(self, threshold: telem.Size) -> None:
        self.threshold = threshold

    def split(self, segment: BinarySegment) -> list[BinarySegment]:
        if segment.size < self.threshold:
            return [segment]
        else:
            truncated = BinarySegment(segment.channel, segment.start,
                                      segment.data[:self.threshold])
            _next = BinarySegment(segment.channel, truncated.end,
                                  segment.data[self.threshold:])
            return [truncated, *self.split(_next)]

File: client/ts/src/lib/segment/splitter.ts
import { ValidationError } from '../errors';
import { Size } from '../telem';

import TypedSegment from './typed';

export default class Splitter {
  threshold: Size;

  constructor(threshold: Size) {
    this.threshold = threshold;
  }

  split(segment: TypedSegment): TypedSegment[] {
    if (segment.size.smallerThan(this.threshold)) return [segment];
    if (!segment.channel.density)
      throw new ValidationError(
        'Cannot split segment because channel density is undefined'
      );
    const splitPoint =
      this.threshold.valueOf() -
      (this.threshold.valueOf() % segment.channel.density.valueOf());
    const truncated = new TypedSegment(segment.channel, {
      ...segment.payload,
      data: segment.payload.data.slice(0, splitPoint),
    });
    const next = new TypedSegment(segment.channel, {
      ...segment.payload,
      data: segment.payload.data.slice(splitPoint),
    });
    return [truncated, ...this.split(next)];
  }
}
